### DISCUSSION TITLE
Stable interface v1

### DISCUSSION THREAD

**Discussion by marco-mariotti** [biocore-ntnu/pyranges#357] [General]
Related to https://github.com/pyranges/pyranges/issues/353 : let's plan a v1 release a newly designed method interface. We'll need to break some backward compatibility, so let's open this discussion with the pyranges community to get feedback.

Among other things, we'll make a major transition in data structure. Currently, PyRanges is a collection of DataFrames. We'll change it to be a a subclass of DataFrame. This brings many advantages but compels to change a few methods to avoid conflicts.

In general, all methods of the future PyRanges object, including those native of pyranges and those inherited from pandas DataFrame, will return a PyRanges object whenever they can -- that is to say, whenever their returned value is a DataFrame with columns Chromosome, Start, End.

The first item of discussion is what to do with conflicts in the namespace of PyRanges and DataFrame.

# Namespace conflicts
The methods/properties below are in both the current PyRanges and DataFrame namespaces. I write my proposal for resolution of conflicts. 

## Identical behavior, no conflict in sight; can safely drop pyranges version
-columns	identical
-copy		identical
-dtypes		identical
-empty		identical
-head		identical
-tail		identical

## Deprecated due to transition from old data structure (collection of dataframes) to new (one dataframe):

-apply	conflict
		pyranges version can be dropped, in favor of gr.groupby('Chromosome').apply or gr.groupby(['Chromosome', 'Strand']).apply
		or should be build a new methods that wraps for that? I vote no
		futurewarning on its use
		
-items		conflict
		pyranges version can be dropped
		futurewarning on its use

-keys		conflict
		pyranges version can be dropped
		futurewarning on its use

-values		conflict
		pyranges version can be dropped
		futurewarning on its use

## With an almost identical dataframe version:
-to_csv		small conflict
		pd offers more args, but not chain=True. Must drop pyranges version
		futurewarning on problematic use only

-drop		small conflict
		pyranges version can be dropped
		futurewarning on use of "like" argument

-assign		small conflict
		pyranges version can be dropped
		futurewarning on its use, since syntax is different from pd

-insert		small conflict
		pyranges version can be dropped
		futurewarning on its use, since syntax is different from pd

-sample		small conflict
		behavior is identical, but pd has more args, and arg order is not respected. Must drop pyranges version
		futurewarning on problematic use only

## Methods related to overlap, to be renamed:
-join		conflict
		rename pyranges join to overlap_join
		futurewarning on its use

-merge		conflict
		rename pyranges join to overlap_merge
		futurewarning on its use
	
-subtract	conflict
		rename pyranges join to overlap_substract
		futurewarning on its use

## Magic methods including get and set attributes
Require extensive discussion; later

**Comment by marco-mariotti** [comment]
# Set and get attributes

In v1, for compatibility with the superclass DataFrame, we need to review these methods.
For simplicity, I will refer here to 'Chromosome', 'Start', 'End' (and optionally 'Strand') as core columns.
In the old pyranges, you could index/access data like this:

```
p=pr.data.chipseq() # PyRanges
df=p.df	            # DataFrame

```

## 1) Get attr with dot operator
```
p.Start    # --> get a core column as Series
p.Score    # --> get a metadata column as Series
```

This is supported by DataFrame; no change required.

## 2) Get row selection with a boolean series indexer

```
selector = (df.Strand == '-')
df [selector]  # --> get a pyranges with rows whose strand is -, or whatever other condition was specified

```
This is supported by DataFrame; no change required.

## 3) Get row selection via regional retrieval

```
p['chrY']  # --> get intervals  whose Chromosome is chrY
p['chrY', 5_000_000:8_000_000]  # only intervals in chrY and included in the a position range
p['chrY', '-', 5_000_000:8_000_000]  # only intervals in chrY, - strand, included in the a position range

```

Unfortunately this must be dropped as it conflicts with Dataframe   df[colname] to get a column.
But we can provide this valuable functionality as a new indexer:

```
newp.loci['chrY']
newp.loci['chrY', 5_000_000:8_000_000] 
newp.loci['chrY', '-', 5_000_000:8_000_000]
```

We can implement this early, in v.0.0.*, issue a futurewarning, then maintain regional retrieval only through this indexer in v1.

I used .chr as name; but other proposals are welcome. Perhaps .gloc? .genloc? .reg? .prloc?



## 4) Get attr by indexing with list of column names
`p[ ['Name', 'Score'] ]   # --> get a pyranges including core columns and the requested Name, Score
`
Instead, this syntax in dataframes will return only the requested columns:

`df[ ['Name', 'Score'] ]   # --> get a dataframe including only Name, Score 
`
I propose to adopt DataFrame syntax. But we may supply the current pyrange functionality as a new function, or a new indexer. 
So we may implement these alternatives:

```
newp.get_with_loc_columns( ['Name', 'Score'] ) # --> get a pyranges including core columns and Name, Score
# newp.prloc[ :, ['Name', 'Score'] ]    # --> get a pyranges including core columns and Name, Score
```
We implemented the first method above.


## 5) Set attr

Old pyranges supported these two syntaxes for setting new columns.

`p=p.assign('newcolumn',  some_fn_returning_value)
`
Which can easily be supported, though with slighlty different syntax, by DataFrame.assign;  and:

`p.newcolumn = some_value
`
This is more problematic. 
The same syntax applied to dataframes will create an attribute, not a new column.  Case in point:


```
df.something=False
df

     Chromosome      Start        End Name  Score Strand
0          chr1  212609534  212609559   U0      0      +
1          chr1  169887529  169887554   U0      0      +
...         ...        ...        ...  ...    ...    ...
9998       chrY    8010951    8010976   U0      0      -
9999       chrY    7405376    7405401   U0      0      -

[10000 rows x 6 columns]
```

```
print(df.something)
False

```

This is how you create a column in pandas:

```
df['newcol']=True
df

     Chromosome      Start        End Name  Score Strand  newcol
0          chr1  212609534  212609559   U0      0      +    True
1          chr1  169887529  169887554   U0      0      +    True
...         ...        ...        ...  ...    ...    ...     ...
9998       chrY    8010951    8010976   U0      0      -    True
9999       chrY    7405376    7405401   U0      0      -    True

[10000 rows x 7 columns]
```


My proposal is to add support for the p['newcol']=value  syntax early, in v.0.0.*, issue a futurewarning for any use as p.newcol; from v1, stick to DataFrame behavior.


## 6) Others syntaxes
While old pyranges did not support any more syntaxes for getattr and setattr (if I haven't forgot anything!), inheritance from DataFrame in v1 will bring plenty, including support for the .index, a big absentee of old pyranges. 
So, all of these will work for getting and setting values: .at .iat .loc .iloc

These may be combined with .chr to set attributes only in certain regions:

```
newp.label=False
new.loc[ newp.loci['chrY', 5_000_000:8_000_000].index,  'label']=True
```

**Comment by marco-mariotti** [comment]
# Roadmap
Here's a google docs with updated roadmap for what to do with each method. This will reflect any accepted suggestions we may receive:

https://docs.google.com/document/d/1zjv-uUBaiccSHQa-lef9H7_mtT7IuWtdKzRSvR00ZPQ/edit?usp=sharing

**Comment by endrebak** [comment]
Possibly of interest: https://peps.python.org/pep-0702/

**Reply by marco-mariotti** [reply]
Nice feature! Problem is, it's coming only in python 3.13. I hope it'll be available as future statement. Or perhaps, copy the decorator. We'll keep an eye out!
